Source/WebCore/rendering/svg/SVGInlineTextBox.cpp

```diff
@@ -554,24 +616,68 @@ void SVGInlineTextBox::paintTextWithShadows(GraphicsContext& context, const Rend
     FloatRect shadowRect(FloatPoint(textOrigin.x(), textOrigin.y() - scaledFont.metricsOfPrimaryFont().floatAscent()), textSize);
 
     GraphicsContext* usedContext = &context;
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    SVGPaintServerHandling paintServerHandling { context };
+#endif
+
+    auto prepareGraphicsContext = [&]() -> bool {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+        if (renderer().document().settings().layerBasedSVGEngineEnabled())
+            return acquirePaintingResource(paintServerHandling, scalingFactor, parent()->renderer(), style);
+#endif
+
+        return acquireLegacyPaintingResource(usedContext, scalingFactor, parent()->renderer(), style);
+    };
+
+    auto restoreGraphicsContext = [&]() {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+        if (renderer().document().settings().layerBasedSVGEngineEnabled()) {
+            releasePaintingResource(paintServerHandling);
+            return;
+        }
+#endif
+
+        releaseLegacyPaintingResource(usedContext, /* path */nullptr);
+    };
+
     do {
-        if (!prepareGraphicsContextForTextPainting(usedContext, scalingFactor, style))
+        if (!prepareGraphicsContext())
             break;
 
         {
             ShadowApplier shadowApplier(style, *usedContext, shadow, nullptr, shadowRect);
 
             if (!shadowApplier.didSaveContext())
                 usedContext->save();
-            usedContext->scale(1 / scalingFactor);
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+            // Optimized code path to support gradient fill/stroke on text without using temporary ImageBuffers / masking.
+            auto* fillGradient = paintingResourceMode().contains(RenderSVGResourceMode::ApplyToFill) ? usedContext->fillGradient() : nullptr;
+            auto* strokeGradient = paintingResourceMode().contains(RenderSVGResourceMode::ApplyToStroke) ? usedContext->strokeGradient() : nullptr;
+            if (renderer().document().settings().layerBasedSVGEngineEnabled() && (fillGradient || strokeGradient)) {
```

Comment:
I think it may work better to set fillGradient/strokeGradient to nullptr and only try assign it a value if layerBasedSVGEngineEnabled() is true. Then below we do not need to check layerBasedSVGEngineEnabled() a second time.

Reply:
Good idea, fixed.